onnxruntime/core/providers/coreml/builders/impl/pad_op_builder.cc

```diff
@@ -0,0 +1,178 @@
+  width_border->set_startedgesize(pads_span[input_rank - 1]);
+  width_border->set_endedgesize(pads_span[2 * input_rank - 1]);
+
+  *layer->mutable_input()->Add() = input_defs[0]->Name();
+  *layer->mutable_output()->Add() = node.OutputDefs()[0]->Name();
+
+  model_builder.AddLayer(std::move(layer));
+
+  return Status::OK();
+}
+#endif
+
+// Operator support related
+
+bool PadOpBuilder::IsOpSupportedImpl(const Node& node, const OpBuilderInputParams& input_params,
+                                     const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+  const auto& initializers = input_params.graph_viewer.GetAllInitializedTensors();
+
+  std::vector<int64_t> input_shape;
+  if (!GetShape(*input_defs[0], input_shape, logger))
+    return false;
+
+  if (input_shape.empty() || input_shape.size() < 2) {
+    LOGS(logger, VERBOSE) << "Pad requires input shape to be at least 2, input is "
+                          << input_shape.size() << "d shape";
+    return false;
+  }
+
+  if (std::find(input_shape.begin(), input_shape.end(), int64_t{0}) != input_shape.end()) {
+    LOGS(logger, VERBOSE) << "Pad input with zero elements for dimension is not supported";
+    return false;
+  }
+
+  {
+    NodeAttrHelper helper(node);
+    const auto mode = helper.Get("mode", "constant");
+    if (mode != "constant") {
+      LOGS(logger, VERBOSE) << "Only support constant mode Pad operator for now, mode: " << mode;
+      return false;
+    }
+
+    if (input_defs.size() < 3) {
+      LOGS(logger, VERBOSE) << "`constant_value` input is required for constant mode Pad op.";
+      return false;
+    }
+  }
+
+  // only support if `pads` input is known and does not contain negative values and only has non-zero padding values
+  // for last two dimensions.
+  {
+    const auto pads_initializer_it = initializers.find(input_defs[1]->Name());
+    if (pads_initializer_it == initializers.end()) {
+      LOGS(logger, VERBOSE) << "pads must be known";
+      return false;
+    }
+
+    const ONNX_NAMESPACE::TensorProto& pads_initializer = *pads_initializer_it->second;
+    Initializer unpacked_tensor(pads_initializer);
+
+    auto tensor_data = unpacked_tensor.DataAsSpan<int64_t>();
+    for (int64_t i = 0; i < unpacked_tensor.size(); i++) {
+      if (tensor_data[i] < 0) {
+        LOGS(logger, VERBOSE) << "Negative pad value is not supported: pads["
+                              << i << "] = " << tensor_data[i];
+        return false;
+      }
+    }
+
+    // Check that the input pads value only have non-zero values on last two dimensions - [H,W].
+    // As CoreML PaddinglayerParams only apply padding on the last two dimensions:
+    // https://apple.github.io/coremltools/mlmodel/Format/NeuralNetwork.html#paddinglayerparams
+    const auto input_rank = SafeInt<int64_t>(input_shape.size());
+    for (int64_t i = 0; i < unpacked_tensor.size(); i++) {
+      if (!(i == input_rank - 1 || i == input_rank - 2 ||
```

Comment:
it should be the number of pad axes instead of input rank

Reply:
If axes is not specified it might be easier to create the default value before applying the check. 

Does the check needs to be that pads can only contain non-zero values for entries in axes that match the last two dims of the input data?

axes could potentially have an entry for only one of the last two dims in the input data so we need to allow for that.

Reply:
> it should be the number of pad axes instead of input rank

sorry, my previous comment was not right. the pad axes can be specified in an arbitrary order.

> Does the check needs to be that pads can only contain non-zero values for entries in axes that match the last two dims of the input data?

yes, I think that is it.

Reply:
I rewrite the check a bit. not sure if it looks the simpliest, but hopefully should include all the cases discussed above. including allowing the case for axes having one entry of the last 2D, etc.

Basically the logic looks like:

```
for (i = 0; i < num_axis; i++)
  if  (axes[i] == r-2)
     padding_coreml[0] = pads[i]
     padding_coreml[2] = pads[i + num_axis]
  if  (axes[i] == r-1) 
     padding_coreml[1] = pads[i]
     padding_coreml[3] = pads[i + num_axis]
  else  //   axes[i] < r - 2
     if pads[i] != 0 or pads[i+ num_axis] != 0
         Not supported
 ```
assuming if axes not provided, we will create a default one as [0,1,2,...,input_rank - 1]